In [16]:
0.4**5

0.010240000000000003

# ACCY 577 Final Project  

## Team Members
Please list your team members in this cell.

----



## Overview 


After several years of operation, the Lending Club&trade; wants to understand how the individual components (i.e., features) of a loan, such as the amount loaned, the term length, or the interest rate, affect the profitability of a specific loan. In particular, they're interested in understanding how likely a loan is to be repaid. In this project, you will use the provided loans dataset to help the Lending Club&trade; figure out which loans are most profitable.


## Criteria

You will work in groups to analyze the data to make recommendations based on the `lending_club_2007_2011_6_states.csv` dataset. 

You will complete three tasks for this group project:
1. A group report in the form of a Jupyter notebook,
2. A video presentation where your group will present your results, and 
3. Peer evaluation of the contributions of each member of your group.

Your final group report will be a single Jupyter notebook that will integrate Markdown, Python code, and the results from your code, such as data visualizations. Markdown cells should be used to explain any decisions you make regarding the data, to discuss any plots or visualizations generated in your notebook, and the results of your analysis. As a general guideline, the content should be written in a way that a fellow classmate (or an arbitrary data scientist/analyst) should be able to read your report and understand the results, implications, and processes that you followed to achieve your result.

All group members should present in the video presentation. You can use presentation software such as MS PowerPoint. The presentations should cover the steps in your analytics process and highlight your results. You don't need to explain Python code in the presentation. Focus on your analysis method and results. The presentation should take between eight to twelve minutes.

### Rubric
  - Notebook report (160 points)
  - Video presentation (100 points)
  - Peer assessment for your groupmates (20 points) - You will get this 20 points as long as you submit peer evaluation before the deadline.
  - Inidividual grade will be adjusted based on your peer evaluations by your group members.
  
### Submissions
- Each group:
 - Report in .ipynb format
 - Report in .html format
 - Presentation link
- Each individual:
 - Peer evaluation

### General

Your report should 
  - Use proper markdown.  
  - Include all of the code used for your analysis.
  - Plots should be properly labeled (e.g., axis labels and titles).
  - Use a consistent style.
  
Your presentation should
  - Use clear, concise slides.
  - Support your points with facts (e.g., statistics, visualizations).
  - Avoid any coding details.
  - Be creative.
-----

## Get Started

In [1]:
%matplotlib inline

import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

#display all dataframe columns in df.head()
pd.options.display.max_columns = None
#display long string in datafame
pd.options.display.max_colwidth = 300

#filter out warning messages
import warnings
warnings.filterwarnings('ignore')

### Business Understanding

#### Lending Club

LendingClub is an American peer-to-peer lending company, headquartered in San Francisco, California. It is the world's largest peer-to-peer lending platform.

LendingClub enables borrowers to create unsecured personal loans between $1,000 and $40,000. Investors can search and browse the loan listings on the LendingClub website and select loans that they want to invest in based on the information supplied about the borrower, amount of loan, loan grade, and loan purpose. Investors make money from interest. LendingClub makes money by charging borrowers an origination fee and investors a service fee.

For more information about the company please check out the Wikipedia article about the [LendingClub](https://en.wikipedia.org/wiki/LendingClub).


### Load Data and Data Dictionary

In [2]:
data_dict = pd.read_csv('data_dictionary.csv')
data_dict

ColumnName  \
0                acc_now_delinq   
1                    addr_state   
2                    annual_inc   
3              application_type   
4      chargeoff_within_12_mths   
5       collection_recovery_fee   
6    collections_12_mths_ex_med   
7          debt_settlement_flag   
8     debt_settlement_flag_date   
9                   delinq_2yrs   
10                  delinq_amnt   
11          disbursement_method   
12                          dti   
13             earliest_cr_line   
14                   emp_length   
15                    emp_title   
16                  funded_amnt   
17              funded_amnt_inv   
18                        grade   
19                hardship_flag   
20               home_ownership   
21          initial_list_status   
22               inq_last_6mths   
23                  installment   
24                     int_rate   
25                      issue_d   
26           last_credit_pull_d   
27              last_pymnt_amnt   
28                 last_pymnt_d   
29                    loan_amnt   
30                  loan_status   
31       mths_since_last_delinq   
32  mths_since_last_major_derog   
33       mths_since_last_record   
34                 next_pymnt_d   
35                     open_acc   
36                    out_prncp   
37                out_prncp_inv   
38                  policy_code   
39                      pub_rec   
40         pub_rec_bankruptcies   
41                      purpose   
42                   pymnt_plan   
43                   recoveries   
44                    revol_bal   
45                   revol_util   
46                    sub_grade   
47                    tax_liens   
48                         term   
49                        title   
50                    total_acc   
51                  total_pymnt   
52              total_pymnt_inv   
53                total_rec_int   
54           total_rec_late_fee   
55              total_rec_prncp   
56          verification_status   
57                     zip_code   

                                                                                                                                                                                                 Description  
0                                                                                                                                            The number of accounts on which the borrower is now delinquent.  
1                                                                                                                                                 The state provided by the borrower in the loan application  
2                                                                                                                              The self-reported annual income provided by the borrower during registration.  
3                                                                                                       Indicates whether the loan is an individual application or a joint application with two co-borrowers  
4                                                                                                                                                                     Number of charge-offs within 12 months  
5                                                                                                                                                                             post charge off collection fee  
6                                                                                                                                           Number of collections in 12 months excluding medical collections  
7                                                                                                         Flags whether or not the borrower, who has charged-off, is working with a debt-settlement company.  
8                                                                                           

In [3]:
loan_df = pd.read_csv('lending_club_2007_2011_6_states.csv')
loan_df.sample(5)

loan_amnt  funded_amnt  funded_amnt_inv       term  int_rate  \
4245       14000        14000      14000.00000  36 months     10.99   
2624       14000        14000      14000.00000  60 months     15.96   
7655        3000         3000       3000.00000  36 months     15.62   
15744       3000         3000       3000.00000  36 months      7.88   
19224       3900         3900       3746.41821  36 months     10.51   

       installment grade sub_grade                  emp_title emp_length  \
4245        458.28     B        B3       One Technologies, LP    2 years   
2624        340.16     C        C5   Florida Power and Light     6 years   
7655        104.91     D        D1        Leonard Carder, LLP    7 years   
15744        93.85     A        A5       American Ingredients    3 years   
19224       126.78     B        B2  Michelle Kaufmann Designs     1 year   

      home_ownership  annual_inc verification_status   issue_d loan_status  \
4245        MORTGAGE    125000.0        Not Verified  Sep-2011  Fully Paid   
2624        MORTGAGE     70000.0     Source Verified  Oct-2011  Fully Paid   
7655        MORTGAGE     61000.0     Source Verified  May-2011  Fully Paid   
15744           RENT     36000.0        Not Verified  Mar-2010  Fully Paid   
19224       MORTGAGE     75000.0        Not Verified  Nov-2008  Fully Paid   

      pymnt_plan             purpose                              title  \
4245           n    home_improvement                   Home Improvement   
2624           n               house                           Personal   
7655           n  debt_consolidation                 Debt Consolidation   
15744          n  debt_consolidation  Graduate School Tuition and Books   
19224          n         credit_card                No more credit card   

      zip_code addr_state    dti  delinq_2yrs earliest_cr_line  \
4245     750xx         TX  17.40          2.0         Dec-1997   
2624     333xx         FL   6.89          0.0         Aug-1998   
7655     953xx         CA  22.78          0.0         Jan-2002   
15744    913xx         CA   9.20          0.0         Aug-1989   
19224    941xx         CA  20.23          0.0         Mar-1996   

       inq_last_6mths  mths_since_last_delinq  mths_since_last_record  \
4245              0.0                    10.0                     NaN   
2624              2.0                    54.0                     NaN   
7655              0.0                     NaN                    54.0   
15744             0.0                     NaN                     NaN   
19224             0.0                    30.0                     NaN   

       open_acc  pub_rec  revol_bal  revol_util  total_acc  \
4245        4.0      0.0          0       25.33       22.0   
2624       15.0      0.0       2608       21.00       32.0   
7655        6.0      1.0      14154       95.00       20.0   
15744       4.0      0.0      16251       83.30       13.0   
19224       6.0      0.0       4508       30.70       15.0   

      initial_list_status  out_prncp  out_prncp_inv   total_pymnt  \
4245                    f        0.0            0.0  15891.858615   
2624                    f        0.0            0.0  20409.180027   
7655                    f        0.0            0.0   3730.564175   
15744                   f        0.0            0.0   3378.401107   
19224                   f        0.0            0.0   4541.212933   

       total_pymnt_inv  total_rec_prncp  total_rec_int  total_rec_late_fee  \
4245          15891.86          14000.0        1891.86                 0.0   
2624          20409.18          14000.0        6409.18                 0.0   
7655           3730.56           3000.0         730.56                 0.0   
15744          3378.40           3000.0         378.40                 0.0   
19224          4361.69           3900.0         641.21                 0.0   

       recoveries  collection_recovery_fee last_pymnt_d  last_pymnt_amnt  \
4245          0.0                 

In [4]:
loan_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 19908 entries, 0 to 19907
Data columns (total 58 columns):
loan_amnt                      19908 non-null int64
funded_amnt                    19908 non-null int64
funded_amnt_inv                19908 non-null float64
term                           19908 non-null object
int_rate                       19908 non-null float64
installment                    19908 non-null float64
grade                          19908 non-null object
sub_grade                      19908 non-null object
emp_title                      18723 non-null object
emp_length                     19409 non-null object
home_ownership                 19908 non-null object
annual_inc                     19908 non-null float64
verification_status            19908 non-null object
issue_d                        19908 non-null object
loan_status                    19908 non-null object
pymnt_plan                     19908 non-null object
purpose                        19908 non-

In [5]:
loan_df.describe()

loan_amnt   funded_amnt  funded_amnt_inv      int_rate  \
count  19908.000000  19908.000000     19908.000000  19908.000000   
mean   11353.846444  11065.763763     10500.929748     12.089717   
std     7463.700492   7176.276661      7106.229650      3.698287   
min      500.000000    500.000000         0.000000      5.420000   
25%     5750.000000   5600.000000      5000.000000      9.450000   
50%    10000.000000  10000.000000      9000.000000     11.860000   
75%    15000.000000  15000.000000     14500.000000     14.610000   
max    35000.000000  35000.000000     35000.000000     24.400000   

        installment    annual_inc           dti   delinq_2yrs  inq_last_6mths  \
count  19908.000000  1.990800e+04  19908.000000  19908.000000    19908.000000   
mean     330.614254  7.107345e+04     13.008619      0.143962        0.829466   
std      210.557434  6.980565e+04      6.663658      0.489576        1.044978   
min       15.690000  4.000000e+03      0.000000      0.000000        0.000000   
25%      171.287500  4.200000e+04      7.840000      0.000000        0.000000   
50%      285.780000  6.000000e+04     13.075000      0.000000        0.000000   
75%      440.815000  8.500000e+04     18.200000      0.000000        1.000000   
max     1302.690000  6.000000e+06     29.990000     11.000000        8.000000   

       mths_since_last_delinq  mths_since_last_record      open_acc  \
count             6998.000000             1209.000000  19908.000000   
mean                35.841097               69.354839      9.278782   
std                 21.599170               44.520279      4.414903   
min                  0.000000                0.000000      2.000000   
25%                 19.000000                0.000000      6.000000   
50%                 34.000000               90.000000      9.000000   
75%                 51.000000              104.000000     12.000000   
max                106.000000              129.000000     44.000000   

            pub_rec      revol_bal    revol_util     total_acc  out_prncp  \
count  19908.000000   19908.000000  19889.000000  19908.000000    19908.0   
mean       0.046715   13363.994826     49.424966     21.527627        0.0   
std        0.217827   15943.303849     28.198395     11.269006        0.0   
min        0.000000       0.000000      0.000000      2.000000        0.0   
25%        0.000000    3770.500000     26.500000     13.000000        0.0   
50%        0.000000    8876.500000     50.300000     20.000000        0.0   
75%        0.000000   16952.250000     72.800000     28.000000        0.0   
max        3.000000  148829.000000     99.900000     90.000000        0.0   

       out_prncp_inv   total_pymnt  total_pymnt_inv  total_rec_prncp  \
count        19908.0  19908.000000     19908.000000     19908.000000   
mean             0.0  12286.852391     11690.155107      9913.519990   
std              0.0   9098.847567      8984.722396      7126.424699   
min              0.0      0.000000         0.000000         0.000000   
25%              0.0   5670.592545      5239.857500      4800.000000   
50%              0.0  10042.735817      9427.845000      8000.000000   
75%              0.0  16681.688450     15983.440000     14000.000000   
max              0.0  58480.139915     58438.370000     35000.020000   

       total_rec_int  total_rec_late_fee    recoveries  \
count   19908.000000        19908.000000  19908.000000   
mean     2277.120129            1.489973     94.722382   
std      2583.733214            7.964257    666.434583   
min         0.000000            0.000000      0.000000   
25%       687.595000            0.000000      0.000000   
50%      1389.230000            0.000000      0.000000   
75%      2842.795000            0.000000      0.000000   
max     23480.140000          180.200000  29623.350000   

       collection_recovery_fee  last_pymnt_amnt  next_pymnt_d  \
count             19908.000000     19908.000000           0.0   
mean                 11.861625 

### Paid-Off Rate and Annual Return

The purpose of the project is to identify the most profitable loans. There are two criteria that help you evaluate the loan portfolio profitability: paid off rate and annual return of the portfolio.

#### Paid-off rate
The dataset has loans initiated from 2007 to 2011. All loans have been either fully paid or charged off. So we will create a 'repaid' column by encoding loan status and map Charged Off to 0 and Fully Paid to 1. 

In [6]:
loan_df.loan_status.value_counts()

Fully Paid     16965
Charged Off     2943
Name: loan_status, dtype: int64

In [7]:
mapping_dict = {'Charged Off':0, 'Fully Paid':1}
loan_df['repaid'] = loan_df.loan_status.map(mapping_dict)
loan_df.repaid.value_counts()

1    16965
0     2943
Name: repaid, dtype: int64

The average repaid will be the paid-off rate for a loan portfolio. For example, we get overall paid-off rate of all loans in the dataset in the next code cell. About 85% of all loans are paid off.

In [8]:
loan_df.repaid.mean()

0.8521699819168174

#### Annual return

Calculation of loan return is very complicated since the loan is paid by monthly installments. In this project, we simplify the calculation by using the total payment and funded amount. For charged-off loans, total payment collected includes post charged-off recoveries. So we can use following formula to calculate the total return:

$TotalReturn = \frac{Total Payment + recoveries}{Funded Amount} - 1$

    The overall return doesn't reflect loan profitability since loans have different terms. It's more accurate to compare annual returns. There are only two terms in the dataset, 36 months and 60 months. The formula to calculate annual return is:

$Annualized Return = (1+Total Return)^{(1/years)} - 1$. 

For example, if overall return of a 36 month loan is 10%, then annualized return = `(1 + 0.1)**(1/3) - 1` = `0.032`.

Again, this is not the true annualized return on a loan investment. But the goal of this project is to identify loans to invest, so we just need a benchmark to evaluate loan performance.

In the following code cells, we first create total_return and annual_return for each loan, then we define a function to calculate annual return of a loan portfolio. The function calculates annual return for 36 month and 60 month loans in a portfolio separately. **Note:** You don't have to use this function in your project. You can define your own ways to evaluate profitability of a portfolio.

In [9]:
loan_df.term.value_counts()

36 months    14852
60 months     5056
Name: term, dtype: int64

In [10]:
loan_df['total_return'] = (loan_df.total_pymnt + loan_df.recoveries)/loan_df.funded_amnt-1
loan_df['loan_term_year'] = loan_df.term.apply(lambda x:3 if x=='36 months' else 5)
loan_df['annual_return'] = loan_df.apply(lambda x:(1+x.total_return)**(1/x.loan_term_year)-1, axis=1)

In [11]:
loan_df.sample(5)

loan_amnt  funded_amnt  funded_amnt_inv       term  int_rate  \
7608        7600         7600           7600.0  36 months      8.49   
13064       4000         4000           4000.0  36 months     15.95   
15521       9000         9000           8950.0  36 months     14.59   
7661        7200         7200           7175.0  36 months      7.49   
6315       10000        10000           9975.0  36 months      5.42   

       installment grade sub_grade            emp_title emp_length  \
7608        239.88     A        A5             bancroft    9 years   
13064       140.53     D        D4   Frontier Telephone  10+ years   
15521       310.19     D        D1                  NaN        NaN   
7661        223.94     A        A4  BIrd Rock Surf Shop    4 years   
6315        301.60     A        A1     Tops Engineering     1 year   

      home_ownership  annual_inc verification_status   issue_d  loan_status  \
7608            RENT     18000.0            Verified  May-2011  Charged Off   
13064       MORTGAGE     50000.0        Not Verified  Aug-2010   Fully Paid   
15521           RENT     43200.0        Not Verified  Apr-2010   Fully Paid   
7661            RENT     26400.0        Not Verified  May-2011   Fully Paid   
6315            RENT     33000.0        Not Verified  Jul-2011   Fully Paid   

      pymnt_plan             purpose                    title zip_code  \
7608           n            vacation            Vacation Loan    081xx   
13064          n                 car                Dash Loan    146xx   
15521          n  debt_consolidation           Trying To Live    110xx   
7661           n  debt_consolidation  Paying off Credit Cards    921xx   
6315           n  debt_consolidation            consolidation    752xx   

      addr_state    dti  delinq_2yrs earliest_cr_line  inq_last_6mths  \
7608          NJ  13.67          0.0         Jun-2001             3.0   
13064         NY  15.86          0.0         Jul-1990             1.0   
15521         NY  18.11          0.0         May-1996             1.0   
7661          CA   9.14          0.0         Jun-1992             0.0   
6315          TX  20.07          0.0         Dec-2000             0.0   

       mths_since_last_delinq  mths_since_last_record  open_acc  pub_rec  \
7608                      NaN                     NaN       5.0      0.0   
13064                     NaN                     NaN      28.0      0.0   
15521                     NaN                     NaN      11.0      0.0   
7661                      NaN                     NaN       5.0      0.0   
6315                      NaN                     NaN       9.0      0.0   

       revol_bal  revol_util  total_acc initial_list_status  out_prncp  \
7608        4236        17.9       16.0                   f        0.0   
13064       8912        18.7       55.0                   f        0.0   
15521      22427        89.4       17.0                   f        0.0   
7661        7414        20.3        5.0                   f        0.0   
6315        1840        14.2       23.0                   f        0.0   

       out_prncp_inv   total_pymnt  total_pymnt_inv  total_rec_prncp  \
7608             0.0   3440.270000          3440.27          2515.06   
13064            0.0   4994.829943          4994.83          4000.00   
15521            0.0  11167.271300         11105.23          9000.00   
7661             0.0   7490.880000          7464.87          7200.00   
6315             0.0  10743.169241         10716.31         10000.00   

       total_rec_int  total_rec_late_fee  recoveries  collection_recovery_fee  \
7608          617.55           14.941846      292.72                     2.98   
13064         994.83            0.000000        0.00                     0.00   
15521        2167.27            0.000000        0.00                     0.00   
7661          290.88            0.000000        0.00                     0.00   
6315          743.17            0.000000        0.00        

In [12]:
loan_df.loan_term_year.value_counts()

3    14852
5     5056
Name: loan_term_year, dtype: int64

In [13]:
def get_portfolio_annual_return(df):
    '''
    Get annual return of 36 and 60 month loans in the portfolio.
    '''
    annual_return_36, annual_return_60 = 0, 0
    df_36 = df[df.loan_term_year==3]
    if(len(df_36)>0):
        return_36 = (df_36.total_pymnt.sum() + df_36.recoveries.sum())/df_36.funded_amnt.sum()-1
        annual_return_36 = (1+return_36)**(1/3)-1
    df_60 = df[df.loan_term_year==5]
    if(len(df_60)>0):
        return_60 = (df_60.total_pymnt.sum() + df_60.recoveries.sum())/df_60.funded_amnt.sum()-1
        annual_return_60 = (1+return_60)**(1/5)-1
    print (f'36 months loan:{len(df_36)}, Annual return:{round(annual_return_36, 4)}')
    print (f'60 months loan:{len(df_60)}, Annual return:{round(annual_return_60, 4)}')
    return annual_return_36, annual_return_60

We get annual returns of all loans in the dataset with the function in the next code cell. The 36-month loans have about 3.03% annual return and the 60-month loans have about 3.13% annual return.

In [14]:
# Overall return of all loans in the dataset
get_portfolio_annual_return(loan_df)

36 months loan:14852, Annual return:0.0303
60 months loan:5056, Annual return:0.0313


(0.03028789309763802, 0.031295585034884166)

### Confusion Matrix Function

The next code cell defines function that plot the confusion matrix. You may use the function in your model evaluation.

In [15]:
# This method produces a colored heatmap that displays the relationship
# between predicted and actual types from a machine learning method.
def confusion(test, predict, labels, title='Confusion Matrix'):
    '''
        test: true label of test data, must be one dimensional
        predict: predicted label of test data, must be one dimensional
        labels: list of label names, ie: ['positive', 'negative']
        title: plot title
    '''

    bins = len(labels)
    # Make a 2D histogram from the test and result arrays
    pts, xe, ye = np.histogram2d(test, predict, bins)

    # For simplicity we create a new DataFrame
    pd_pts = pd.DataFrame(pts.astype(int), index=labels, columns=labels )
    
    # Display heatmap and add decorations
    hm = sns.heatmap(pd_pts, annot=True, fmt="d")    
    hm.axes.set_title(title, fontsize=20)
    hm.axes.set_xlabel('Predicted', fontsize=18)
    hm.axes.set_ylabel('Actual', fontsize=18)

    return None

----

# Notebook Report Tasks (160 points)

There are four tasks you need to finish for this notebook report. Each task accounts for certain points. There is a  total of 160 points for this notebook report.

- Data understanding and feature selection (60 points)
- Modeling (40 points)
- Construct loan portfolio (40 points)
- Conclusion (20 points)

Please include all needed Python code in this notebook. Each task should have a summary. 

**Please run all code cells before this cell before proceeding.**

## Task 1: Data understanding and feature selection (60 points)

Use domain knowledge and exploratory data analysis to make the initial feature selection. You need to justify why you discard or select a feature. For example, you may say, the following features are excluded since they are only available after a loan is initiated: feature1, feature2, ....

Please include all Python code used to help you make the selection. You can have as may code cells as needed and you will need to summarize your selections and reasons in a markdown cell.

After Task 1, you should have a new DataFrame that contains the selected features. This DataFrame should also include the three features that help to calculate loan returns, total_pymnt, recoveries and funded_amnt. You will not use the three features to select loans, but you will need them to evaluate your model and portfolio.

You may use (but are not limited to) the following criteria to select features:

- Exclude features that are only available after loan initiation. For example, last_pymnt_amnt is not a feature you can use to help you select loans.
- Exclude features that have too many missing values **and** the missing values are hard to fill.
- Exclude categorical features that have too many categories.
- Exclude categorical features that have only one category.
- Exclude features that have just one value.
- Include features that have impact on paid-off rate and return.

You may use(but not limited to) following techniques:

- Domain knowledge(common sense). For example, annual income is likely to have positive impact on paid-off rate.
- Exploratory data analysis, which includes:
    - Descriptive statistics
    - Groupby
    - Pivot table
    - Visualization

You are **not** limited or required to use the techniques listed above.

## Task 2: Modeling (40 points)

Construct a classification model to predict whether a loan will be paid in full. Column 'repaid' will be the label. The purpose of the model is to identify loans that are more likely to be paid in full.

#### Preprocessing

- Separate label and data. Label is 'repaid' column. 
- Encode categorical features
- Manage missing values(drop or fill), please justify your choice.
- Split dataset to train and test. Please set random_state to ensure reproducibility.

#### Modeling, Evaluation and Optimization

- Apply at least two classifiers including LogisticRegression.(The reason is explained in Task 3.)
- If you choose Support Vector Machine, use LinearSVC instead of SVC. The reason is that SVC is very slow with this dataset.
- Apply feature selection with following techniques (you don't have to use all of them):
    - Filter methods
    - Wrapper methods
    - Embedded methods
- Optimize each classifier with cross validation. Select best options for major hyperparameters of the models you choose.
- Evaluate models with classification report and confusion matrix.
- Compare models with ROC plot.

#### Conclusion
Summarize your analysis in Task 2.


## Task 3: Construct Loan Portfolio (40 points)

Construct a loan portfolio out of the test set with the help of your classification model. The portfolio can be a small subset of all loans in the dataset, for example, 50% of all loans in the test set.

For this task, finish the following mini tasks.
- Predict on the test set with optimized model.
- Select loans that are predicted to be paid off by the model.
- Calculate paid-off rate of the selected loans. Is the paid off rate better than the overall paid off rate(85.2%)?
- Calculate annual return of selected loans.
- Assess the risk of your portfolio. You may use the variance of annual return as a risk benchmark.
- Adjust class_weight to 'balanced' and other options (ie. class_weight={1:0.2, 0:0.8}), and use the new model to construct the loan portfolio. How is the accuracy rate change? How is the paid-off rate and return of the new portfolio?
- Discuss other possible ways to construct better loan portfolio. For example, what's the impact of only selecting loans from borrowers that don't have bankruptcy record(pub_rec_bankruptcies==0)?
- Summarize your finding and the performance of your portfolio.

**Note:**  
Some machine learning models are not sensitive to class_weight change. You may change some hyperparameter values to make the model more responsive to class_weight change. For example, set max_depth to 10 for a RandomForestClassifier. This is an advanced topic and you are not required to do it. Out of the models we learned, LogisticRegression is most sensitive to class_weight change. That's why you are asked to include LogisticRegression in your modeling choices in Task 2.

#### Conclusion
Summarsize your analysis in Task 3. 
- Is the classification model helpful in choosing portfolio? 
- Does a portfolio with a higher paid-off rate always have a higher return?
- What other methods can you use to help you choose a portfolio with higher return and lower risk?

## Task 4: Conclusion (20 points)

Summarize:
- Your findings
- The limitations of your analysis
- Possible ways to improve your analysis
- Anything else worth mentioning